onnxruntime/core/optimizer/graph_transformer_mgr.cc

```diff
@@ -2,24 +2,50 @@
 // Licensed under the MIT License.
 
 #include "core/optimizer/graph_transformer_mgr.h"
+#include "core/optimizer/rule_based_graph_transformer.h"
 using namespace onnxruntime;
 using namespace ::onnxruntime::common;
 
 namespace onnxruntime {
 
-Status GraphTransformerManager::ApplyAll(Graph& graph) const {
+common::Status GraphTransformerManager::ApplyTransformers(Graph& graph, std::vector<std::string> providers, const TransformerLevel& level) const {
+  const auto& transformers = level_to_transformer_map_.find(level);
+  if (transformers == level_to_transformer_map_.end()) {
+    return Status::OK();
+  }
+
+  // sort the providers. This makes it easy to find intersection of requested providers for this run
+  // and registered providers for a transformer
+  std::sort(providers.begin(), providers.end());
+
   for (unsigned step = 0; step < steps_; ++step) {
-    bool changed = false;
-    for (auto& transformer : transformers_) {
-      bool t_changed = false;
-      Status s = transformer->Apply(graph, t_changed);
-      if (!s.IsOK()) {
-        return s;
+    bool graph_changed = false;
+    for (const auto& transformer : transformers->second) {
+      const auto compatible_providers = GenerateCompatibleProvidersList(providers, transformer->Name());
```

Comment:
If my initial comment about exec providers and transformers is correct, I think we can avoid this.

Reply:
Yes after our discussion regarding L1 being EP agnostic always... this is going away

---
In reply to: [265762684](https://github.com/Microsoft/onnxruntime/pull/608#discussion_r265762684) [](ancestors = 265762684)